xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,264 @@
+/** @file listnet_ranker.cc
+ *  @brief Implementation of ListNetRanker
+ */
+/* Copyright (C) 2014 Hanxiao Sun
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+/*
+ListNET is adapted from the paper:
+Cao, Zhe, et al. "Learning to rank: from pairwise approach to listwise approach."
+Proceedings of the 24th international conference on Machine learning. ACM, 2007.
+*/
+
+#include "xapian-letor/ranker.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+using namespace Xapian;
+
+ListNETRanker::ListNETRanker() {
+    // default values
+    learning_rate = 0.0001;
+    iterations = 15;
+    cout << "Initializing ListNETRanker with default parameters" << endl;
```

Comment:
You can use the `LOGCALL` macros for this (they're in `common/debuglog.h`) although it may involve linking in another object file. But it's probably a good idea to do a pass and convert all of these at some point.